app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -49,23 +44,35 @@ def xms_mirid_hash
       # according to fields we need to retrieve from the claim.
       # ultimately, transforming "/key1/value1/key2/value2" to {"key1" => "value1", "key2" => "value2"}
       def parsed_xms_mirid
-        split_xms_mirid = @xms_mirid_token_field.split('/')
+        begin
+          split_xms_mirid = @xms_mirid_token_field.split('/')
 
-        if split_xms_mirid.first == ''
-          split_xms_mirid = split_xms_mirid.drop(1)
-        end
+          if split_xms_mirid.first == ''
+            split_xms_mirid = split_xms_mirid.drop(1)
+          end
 
-        index = 0
-        split_xms_mirid.each_with_object({}) do |property, xms_mirid_hash|
-          case property
-          when "subscriptions"
-            xms_mirid_hash["subscriptions"] = split_xms_mirid[index + 1]
-          when "resourcegroups"
-            xms_mirid_hash["resourcegroups"] = split_xms_mirid[index + 1]
-          when "providers"
-            xms_mirid_hash["providers"] = split_xms_mirid[index + 1, index + 3]
+          index = 0
+          split_xms_mirid.each_with_object({}) do |property, xms_mirid_hash|
+            case property
+            when "subscriptions"
+              xms_mirid_hash["subscriptions"] = split_xms_mirid[index + 1]
+            when "resourcegroups"
+              xms_mirid_hash["resourcegroups"] = split_xms_mirid[index + 1]
+            when "providers"
+              xms_mirid_hash["providers"] = split_xms_mirid[index + 1, index + 3]
+            end
+            index += 1
           end
-          index += 1
+        rescue => e
+          raise Err::XmsMiridParseError.new(@xms_mirid_token_field, e.inspect)
+        end
+      end
+
+      def validate_xms_mirid_format
```

Comment:
@sigalsax do you think that we should validate also that the value of `providers` has 3 parts?

Reply:
I don't think we should limit ourselves to 3 parts... regardless, we specify `index + 1, 3` so we will take the 3 elements and otherwise an exception will be caught in try/catch clause above

Reply:
That's not entirely correct. In case of the following xms_mirid:
```
/subscriptions/<subscription-id>/resourcegroups/<some-value>/providers/<just_one_value>
```
then `split_xms_mirid[index + 1, 3]` will be `[<just_one_value>]`. So if the providers part is less than 3 parts then `split_xms_mirid[index + 1, 3]` will return less than 3 parts and the `providers` field will be invalid.

Reply:
More importantly - Why don't you think we should limit ourselves to 3 parts? We already do it in the code.

Reply:
Okay, I was under the assumption that `split_xms_mirid[index + 1, 3]` but throw an index out of bounds exception.

Reply:
So let's check for the length as well so that if Azure changes in the future, the user will know where the failure occured